### ISSUE TITLE
Unable to use DeepOmit with generic type

### ISSUE THREAD

**Issue by xenoterracide** [ts-essentials/ts-essentials#343]
```ts
export type ProducerOptions = DeepOmit<ProducerDocument, EntityOmit>;

export type ProducerDocument = DeepOmit<
  // eslint-disable-next-line @typescript-eslint/ban-types
  OmitProperties<ProducerAggregateRoot, Function>,
  AggregateOmit
>;
```

```
Type 'AggregateOmit' does not satisfy the constraint 'DeepModify<T>'.
  Type 'AggregateOmit' is not assignable to type 'T extends WeakMap<infer K extends object, infer E> ? WeakMap<K, DeepModify<E>> : never'.ts(2344)
```

I'm repeatedly doing this

```ts
export interface Entity<Id> {
  id: Id;
  createdAt: Date;
  modifiedAt: Date;
}

export type EntityOmit = DeepReplace<Entity<never>, never>;
export type AggregateOmit = {
  autoCommit: never;
  [index: symbol]: never;
};
```

all extend [`AggregateRoot`](https://github.com/nestjs/cqrs/blob/master/src/aggregate-root.ts)

this doesn't work though

```ts
export type OmitIt<T extends AggregateRoot> = DeepOmit<
  // eslint-disable-next-line @typescript-eslint/ban-types
  OmitProperties<T, Function>,
  AggregateOmit
>;
```

**Comment by Beraliv** [comment]
Hey!

Let me have a look at it quickly

**Comment by Nightbr** [comment]
Same here, cannot be able to DeepOmit on generic:

```
import { DeepOmit } from 'ts-essentials';

export type ExposedData<T> = DeepOmit<
  T,
  {
    id: never;
    isLatest: never;
    checkId: never;
    createdAt: never;
    updatedAt: never;
  }
>;
```

```
Type '{ id: never; isLatest: never; checkId: never; createdAt: never; updatedAt: never; }' does not satisfy the constraint 'DeepModify<T>'.
  Type '{ id: never; isLatest: never; checkId: never; createdAt: never; updatedAt: never; }' is not assignable to type 'T extends WeakMap<infer K extends object, infer E> ? WeakMap<K, DeepModify<E>> : never'.ts(2344)
```

Here is a TypeScript playground: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgEQKarAeRMeBfOAMyghDgHIYBnAWlSqtQDsZgBDAGyvIG4BYAFCDUAD0iw4MAJ5hUcAKJiIjACYBhABaoAxgGtkbGGwA8AFQB8cALwp0WHDGOC4cUwBpniTy+AqAXHBMqABuqFD8Ai4+VAAyhvQwAUGh4d5w2lp6AJL+gSFhEVHpUKjxKgCCiXkphVEArmAqZZVJ+amRcHiC5hFAA

**Comment by Beraliv** [comment]
Thank you for the playground, I will spend time this week to have a look at it

**Comment by Beraliv** [comment]
There are different approaches that I came up with:

## Another type

`UnsafeDeepOmit` – https://tsplay.dev/WylaKm – analogue for `DeepOmit`, but there's not generic constraint applied

Advantages – works with generic type
Disadvantages – doesn't highlight the typo if any, 2 types instead of one

## Another type + generic constraint

`UnsafeDeepOmit` – https://tsplay.dev/W4XEKW + added generic constraint for object to force to use generic constraint even though it's not used (e.g. `Record<"id" | "isLatest" | "checkId" | "createdAt" | "updatedAt", any>` in this case)

Advantages – works with generic type, highlights the typo if any
Disadvantages – still 2 types

I'm open for the feedback @Nightbr @xenoterracide , let me know about the name and the approaches that I suggested. Do you have other approaches preferable, could you please explain why?

In case there are some significant changes suggested, that's what `Deep`-types currently support:

1. autocomplete for filter keys (with `DeepModify`)
2. filter union support (with `DeepModify`)
3. filtering keys for objects in place (key remapping) for readability

I would want to keep this functionality (and it's covered with tests)

**Comment by Nightbr** [comment]
Thanks a lot for this very detailed exploration. My preference goes to `GenericDeepOmit` and highlight the usage in documentation. 

Add to that, I'm ok that `DeepOmit` doesn't work with not typed generic but maybe we could find a way to make it work when generic extend a defined type.

```ts
import { DeepOmit } from 'ts-essentials';

type GenericCheckData = {
  id: string;
  isLatest: boolean;
  checkId: string;
  createdAt: Date;
  updatedAt: Date;
}

export type ExposedCheckData<T extends GenericCheckData> = DeepOmit<
  T,
  {
    createdAt: never;
    updatedAt: never;
  }
>;

// Result type should be T without createdAt and updatedAt.
```

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgEQKarAeRMeBfOAMyghDgHIYBnAWlSqtQDsZgBDAGyvIG4BYAFCCYATzCo4AcWaoowAMYBhABap5Aa2RsYbOAF5EguHGAATAFxwqMOUwDm-AceBUAMtvoxLAIwgQOqGxMjsbyqhoAkhZWNsD2IXDyUIEwqKYAgl4oHgkArmCmHhlZWqmOeIKCqAAekLBwouJwAKK1EIymKmqa2mwAPAAqcDWpTKZUUjJySuE9OgB8+ijoWDgwfUZwAwA0mwiboclFmZZMqABusgnG+YWpxacXV5sVAvOOQA

Did you explore with the extends on Generic?

**Comment by xenoterracide** [comment]
For the 2 options... what doesn't work with option 2 that would work with option one? like why is option 2 a bad idea. Haven't tested anything yet personally, putting this on my todo list this week.

**Comment by Beraliv** [comment]
I plan to address it after this PR – https://github.com/ts-essentials/ts-essentials/pull/347

I've currently documented the usage and the limitations, it's next on the list, I will let you know when it's done

**Comment by Beraliv** [comment]
To proceed with this enhancement, I'd like to use a following naming convention (that I'd like to document anyway):

- Using `Strict*` prefix when adding a generic constraint (e.g. `StrictOmit` or `StrictExtract`) which don't allow to use the generic type

`Unsafe*` prefix would serve the purpose but I don't like the idea of having unsafe type utility functions because it goes against the idea of `ts-essentials` which is:

> ts-essentials is a set of high-quality, useful TypeScript types that make writing type-safe code easier.

Reference: https://github.com/ts-essentials/ts-essentials?tab=readme-ov-file#api

Therefore I propose the following changes:
- `DeepNonNullable`, `DeepNullable`, `DeepOmit`, `DeepPartial`, `DeepPick`, `DeepReadonly`, `DeepRequired`, `DeepUndefinable` and `DeepWritable` will rename to have a prefix `StrictDeep*` instead (i.e. `StrictDeepNonNullable`, `StrictDeepNullable`, `StrictDeepOmit`, `StrictDeepPartial`, `StrictDeepPick`, `StrictDeepReadonly`, `StrictDeepRequired`, `StrictDeepUndefinable` and `StrictDeepWritable` respectively)
- I will introduce methods without `Strict` prefix that will not have generic constraint therefore will work with generic types

**Comment by Beraliv** [comment]
PR is in progress - https://github.com/ts-essentials/ts-essentials/pull/384, please let me know if you have any input cc @xenoterracide @Nightbr

**Comment by Beraliv** [comment]
> I would say it is risky to introduce a BC in existing helpers.

It wouldn't be a BC as I'm removing a generic constraint on the second type parameter of `DeepOmit` and `DeepPick`. It would be a breaking change if it introduces a new TypeError but instead, it's removing the validating part against a structure of the first parameter type `Type` which doesn't affect existing users.

> Also what about your first suggestion to add `GenericDeepOmit` using the prefix `Generic*` to helpers that supports generics? Would avoid the BC on exisiting helpers.

`Generic*` prefix would be a new convention and will be only semantical. I find it quite misleading because it also works for non-generic cases. So as a user, I would be very confused about what to use so I would need to go to docs and read it anyway.

> Also is `strict` really the right term to describe not working with generic? Generic is a standard way to type complex object with legit polymorphism use cases and I would not say it's not strict.

With regard to `Strict*` prefix, it is used when the additional generic constraint is applied to the utility type parameter, e.g. `StrictOmit<Type, Keys>` adds a generic constraint for `Keys extends keyof Type` to be able to pass only existing keys. It will have the same side effect on generic types which is why it was removed from built-in types in TypeScript. The exact term name is to be clarified. Still, given this term is already used in ts-essentials, it would be consistent for users to continue using it instead of adding another convention on top of it.

**Comment by Nightbr** [comment]
Thanks again for the detailed answer and explanation. Understood the decision and implication, I would still recommend introduce the changes as a major version. In my own use-case, It will work perfectly with our usage (with no BC) so thanks a lot!

**Comment by Beraliv** [comment]
The PR - https://github.com/ts-essentials/ts-essentials/pull/384 was just merged, the changes will be released by the end of April as major

**Comment by Beraliv** [comment]
It's released just now as a part of [v10.0.0](https://github.com/ts-essentials/ts-essentials/releases/tag/v10.0.0)
